### ISSUE TITLE
Percent string literals should allow to escape delimiters

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#5403]
Crystal has no unescaped string literal that can hold really arbitrary strings (that's the essence of @jeff-hykin's [issue](https://github.com/crystal-lang/crystal/issues/501#issuecomment-352479137) (originally posted as comment to #501).

Double-quoted strings are escaped and interpolated so you'd have to do a lot of escaping to use certain characters.
The percent string literals `%q()`, `%q[]`, `%q{}`, `%q<>` and `%q||` allow to write unescaped strings and offer some variaty for delimiters but they are still limited such that you can't use them for a string that contains *all* of these delimiters (or at least one per pair) because the delimiter can't be escaped either.

> Their may be a way I'm unaware of to escape the ) inside of %q().

I don't think there is, but it should probably be escapable as `\)`.

That's also supported in both Ruby and Perl (where this literal syntax was originally borrowed from):
```ruby
%q(\))  # => ")"  (Ruby)
q(\))   # => ")"  (Perl)
```

Having this feature for percent strings in Crystal would be great. It would be relatively easy to implement and shouldn't break anything since escaped percent delimiters are currently invalid code.
And this is certainly better than rededicating single quote delimiters (which is completly infeasible) or introducing an additional delimiter syntax.

**Comment by jeff-hykin** [comment]
Currently heredocs can technically accomplish this, but they can be unwieldy, especially when compared to Ruby's literal string. For those not caught up on #501, Crystal's single quotes are already used for characters, and that issue is already closed.

Supporting an arbitrary literal string (beyond heredoc) would be especially useful for porting code from Ruby.

**Comment by straight-shoota** [comment]
Yes, heredoc is mostly intended for multi-line strings. As far as I can tell, the delimiter can't be escaped as well (but you can make it destinct enough that this shouldn't be a problem). For single-line strings they're impractical anyway.

**Comment by straight-shoota** [comment]
True. Thanks for pointing that out! I've updated the OP.
This issue should be relatively rare and easy to fix, though.

Also affected are nested delimiters prefixed by a backslash. The backslashes will need to be escaped:

```crystal
%q(\)   -> %(\\)
%q(\()) -> %(\\())
%q((\)) -> %((\\))
```

**Comment by FnControlOption** [comment]
Should backslashes always be escaped or should they only be escaped when they precede a nested delimiter? For example, `puts %q(\\\()` in Ruby outputs `\(`, while Crystal currently outputs `\\\(`.

I think I prefer the status quo because `%q` strings in Crystal that contain backlashes are much easier to read:

```sh-session
> rg --only-matching --no-line-number '%q\([^)]*\\\\.*?\)'
spec/std/http/http_spec.cr
%q(foo\"\\bar\ baz\\)
%q(foo\"bar\\baz)

spec/std/process_spec.cr
%q("C:\\\"foo\" project.txt")
%q(C:\\\"foo\"_project.txt)
%q(a\\b\ c\)
%q("a\\b\ c\\")
%q(a\\b\c\)

src/http/common.cr
%q(\"foo\\bar\")
%q(\"foo\\\ bar\")
%q(\"foo\\\ bar\")

src/process/shell.cr
%q("C:\\\"foo\" project.txt")
```

(Also worth considering is how long this issue has been open and how few upvotes it has received.)
